src/main/cpp/src/map_utils.cu

```diff
@@ -0,0 +1,754 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <limits>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/utilities/algorithm.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/io/detail/nested_json.hpp>
+#include <cudf/strings/detail/combine.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/string_view.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+
+//
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+//
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/find.h>
+#include <thrust/for_each.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/permutation_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/pair.h>
+#include <thrust/scan.h>
+#include <thrust/sequence.h>
+#include <thrust/transform.h>
+#include <thrust/transform_reduce.h>
+
+//
+#include <cub/device/device_radix_sort.cuh>
+
+//#define DEBUG_FROM_JSON
+
+#ifdef DEBUG_FROM_JSON
+#include <sstream>
+#endif
+
+namespace spark_rapids_jni {
+
+using namespace cudf::io::json;
+
+namespace {
+
+#ifdef DEBUG_FROM_JSON
+
+// Print the content of the input device vector for debugging.
+template <typename T, typename U = int>
+void print_debug(rmm::device_uvector<T> const &input, std::string const &name,
+                 std::string const &separator, rmm::cuda_stream_view stream) {
+  auto const h_input = cudf::detail::make_host_vector_sync(
+      cudf::device_span<T const>{input.data(), input.size()}, stream);
+  std::stringstream ss;
+  ss << name << ":\n";
+  for (size_t i = 0; i < h_input.size(); ++i) {
+    ss << static_cast<U>(h_input[i]);
+    if (separator.size() > 0 && i + 1 < h_input.size()) {
+      ss << separator;
+    }
+  }
+  std::cerr << ss.str() << std::endl;
+}
+
+// Convert the token value into string name, for debugging purpose.
+std::string token_to_string(PdaTokenT const token_type) {
```

Comment:
I think we should keep all the debug code. They were very helpful for me when working on this. In case there are new bugs uncovered in the future, we can use such debug code to track them down.

Such code will be disabled by default and will never be compiled thus there is nothing to worry about.